### ISSUE TITLE
Use reference types in endpoints

### ISSUE THREAD

**Issue by aturon** [http-rs/tide#3]
Currently, there's no way to write a trait bound that corresponds to an `async fn` that takes borrowed arguments. Because of that restriction, endpoints have to be passed fully owned data, which forces app data to be a cloneable handle.

Once https://github.com/rust-lang/rust/issues/51004 is closed, revisit these APIs (which will also have an impact on the structure of `Middleware`).

**Comment by yoshuawuyts** [comment]
Note from triage: this feature might not even be needed if we end up going with a more layered design in Tide.

**Comment by mmrath** [comment]
@yoshuawuyts Is there any notes on what was discussed on this? 
I would like to enhance the Cookie support such that it is only computed once per request. That computation is then stored in request extension. Any middleware or request handler can then use this Cookie. Also allow middleware/handlers can modify the cookie. All such modifications can then be written to response from a middleware.

**Comment by yoshuawuyts** [comment]
@mmrath if I'm remembering correctly, then by moving the extractors out of Tide's core we can simplify the design enough that this is no longer needed.

The extractors would then become a concept _on top of_ the core API. The exact semantics around that haven't been decided yet, as nailing the core API is a bit more of a priority at the moment.

Does that more or less answer it?

**Comment by mmrath** [comment]
> Does that more or less answer it?

Yes, somewhat. I will wait for the core api design.

**Comment by yoshuawuyts** [comment]
@mmrath speaking of which: https://github.com/rustasync/tide/pull/156 is now up! :tada:

**Comment by aturon** [comment]
Update after #156:

To clarify: we'll still be able to make some API improvements once this compiler limitation is lifted. In particular, right now middleware has to pass a fully *owned* context downward, which means that if you want to read information out of the context afterward, you have to arrange for that separately.

Ultimately, both middleware and endpoints should work with `&mut Context` once this issue is unblocked.

**Comment by yoshuawuyts** [comment]
We've moved to only taking owned values in endpoints and that seems to be working alright for us. Once we have GATs we may be able to reconsider, but for now it seems fine. Going to go ahead and close this.
